internal/provider/resource_access_model.go

```diff
@@ -5,186 +5,303 @@ import (
+					stringvalidator.OneOf("admin", "read", "write"),
 				},
 			},
-			"access": {
-				Description:  "Type of access to the model",
-				ValidateFunc: validation.StringInSlice([]string{"admin", "read", "write"}, false),
-				Type:         schema.TypeString,
-				ForceNew:     true,
-				Required:     true,
+			// ID required by the testing framework
+			"id": schema.StringAttribute{
+				Computed: true,
 			},
 		},
 	}
 }
 
-func resourceAccessModelCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	client := meta.(*juju.Client)
+// Configure enables provider-level data or clients to be set in the
+// provider-defined DataSource type. It is separately executed for each
+// ReadDataSource RPC.
+func (a *accessModelResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
 
-	var diags diag.Diagnostics
+	client, ok := req.ProviderData.(*juju.Client)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *juju.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
+	a.client = client
+}
 
-	model := d.Get("model").(string)
-	access := d.Get("access").(string)
-	usersInterface := d.Get("users").([]interface{})
-	users := make([]string, len(usersInterface))
-	for i, v := range usersInterface {
-		users[i] = v.(string)
+func (a *accessModelResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan accessModelResourceModel
+
+	// Read Terraform configuration from the request into the model
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
-	uuid, err := client.Models.ResolveModelUUID(model)
+	// Get the users
+	var users []string
+	plan.Users.ElementsAs(ctx, users, false)
+
+	modelNameStr := plan.Model.ValueString()
+	// Get the modelUUID to call Models.GrantModel
+	uuid, err := a.client.Models.ResolveModelUUID(modelNameStr)
 	if err != nil {
-		return diag.FromErr(err)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create access model resource, got error: %s", err))
+		return
 	}
-
 	modelUUIDs := []string{uuid}
 
+	accessStr := plan.Access.ValueString()
+	// Call Models.GrantModel
 	for _, user := range users {
-		err := client.Models.GrantModel(juju.GrantModelInput{
+		err := a.client.Models.GrantModel(juju.GrantModelInput{
 			User:       user,
-			Access:     access,
+			Access:     accessStr,
 			ModelUUIDs: modelUUIDs,
 		})
 		if err != nil {
-			return diag.FromErr(err)
+			resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create access model resource, got error: %s", err))
+			return
 		}
 	}
+	plan.ID = types.StringValue(fmt.Sprintf("%s:%s:%s", modelNameStr, accessStr, strings.Join(users, ",")))
```

Comment:
During qa, the ID after creation was missing the user name:
```
juju_access_model.this: Creation complete after 0s [id=machinetest:write:]
```
If I run `terraform plan` again, it wants to add a user heather
```
  # juju_access_model.this will be updated in-place
  ~ resource "juju_access_model" "this" {
      ~ id     = "machinetest:write:" -> (known after apply)
      + users  = [
          + "heather",
        ]
        # (2 unchanged attributes hidden)
    }
```
Happens for creation only, import is good.

Reply:
Is this still  true? I can't seem to  reproduce this.

Reply:
i can no longer reproduce